app/controllers/foreman_tasks/api/tasks_controller.rb

```diff
@@ -149,23 +149,39 @@ def index
         }
       end
 
-      api :POST, '/tasks/callback', N_('Send data to the task from external executor (such as smart_proxy_dynflow)')
-      param :callback, Hash do
-        param :task_id, :identifier, :desc => N_('UUID of the task')
-        param :step_id, String, :desc => N_('The ID of the step inside the execution plan to send the event to')
+      def_param_group :callback_target do
+        param :callback, Hash do
+          param :task_id, :identifier, :desc => N_('UUID of the task')
+          param :step_id, String, :desc => N_('The ID of the step inside the execution plan to send the event to')
+        end
+      end
+
+      def_param_group :callback do
+        param_group :callback_target
+        param :data, Hash, :desc => N_('Data to be sent to the action')
       end
-      param :data, Hash, :desc => N_('Data to be sent to the action')
+
+      api :POST, '/tasks/callback', N_('Send data to the task from external executor (such as smart_proxy_dynflow)')
+      param_group :callback
+      param :callbacks, Array, :of => param_group(:callback)
       def callback
-        task = ForemanTasks::Task::DynflowTask.find(params[:callback][:task_id])
-        ForemanTasks.dynflow.world.event(task.external_id,
-                                         params[:callback][:step_id].to_i,
-                                         # We need to call .to_unsafe_h to unwrap the hash from ActionController::Parameters
-                                         ::Actions::ProxyAction::CallbackData.new(params[:data].to_unsafe_h))
+        callbacks = params.key?(:callback) ? Array(params) : params[:callbacks]
+        callbacks.each do |payload|
+          process_callback(payload[:callback][:task_id], payload[:callback][:step_id].to_i, payload[:data].to_unsafe_h)
+        end
         render :json => { :message => 'processing' }.to_json
       end
 
       private
 
+      def process_callback(task_id, step_id, data)
+        task = ForemanTasks::Task::DynflowTask.find(task_id)
+        ForemanTasks.dynflow.world.event(task.external_id,
+                                         step_id,
+                                         # We need to call .to_unsafe_h to unwrap the hash from ActionController::Parameters
```

Comment:
Is this comment still valid? I don't see the usage of `to_unsafe_h` here

Reply:
It is being called [here](https://github.com/theforeman/foreman-tasks/pull/369/files#diff-76e220fa2f3fb075d1aab086606ce317R170), I should move it there